parsing/unit_info.mli

```diff
@@ -0,0 +1,136 @@
+type target_prefix = (string, unit) t
+(** The prefix of a target file with its module name and source file *)
+
+(** Target subfamily *)
+type 'a any_target = ('a, string) t
+
+type target = string any_target
+type target_only = unit any_target
+(** A target file without a source file: this typically happen when packing *)
+
+(** [source_file u] returns the source file of [u] *)
+val source_file: (string,_) t -> string
+
+(** [source_file_opt u] returns the source file of [u] if it exists *)
+val source_file_opt: _ t -> string option
+
+(** [prefix u] returns the prefix of the target. *)
+val prefix: _ t -> string
+
+(** [filename u] returns the file name of the target. *)
+val filename: _ any_target -> string
+
+(** [modname u] returns the module name of the compilation file.*)
+val modname: _ t -> string
+
+(** {1:modname_from_strings Module name computation} *)
+
+(** [modulize s] capitalizes the first letter of [s]. *)
+val modulize: string -> string
+
+(** [modname_from_source filename] is [modulize stem] where [stem] is the
+    basename of the filename [file] stripped from all its extensions.*)
+val modname_from_source: string -> string
+
+(** {2:module_name_validation Module name validation function}*)
+
+(** [is_unit_name ~strict name] returns true only if [name] can be used as a
+    correct module name. The [strict] mode enforces that the first character is
+    an ascii capital letter. *)
+val is_unit_name : strict:bool -> string -> bool
+
+(** [check_unit_name file] prints a warning if the derived module name [modname
+    file] should not be used as a module name. *)
+val check_unit_name : (string,_) t -> unit
+
+(** {1:info_creation Creation of file metadata} *)
+
+(** [from_source filename] associates the module name [modname_from_source
+    filename] to the source file [filename]. *)
+val from_source: string -> source_file
+
+(** [from_target filename] associates the target [filename] to the module name
+    [modname_from_source filename]. *)
+val from_target: string -> target_only
+
+(** [from_target_prefix ~source_file target_prefix] aasociates both the
+    [source_file] and the module name [modname_from_source target_prefix] to
+    [target_prefix].
+
+   This function emits a warning if the derived module name is not valid
+   according to {!is_unit_name}[ ~strict:true].
+*)
+val from_target_prefix: source_file:string -> string -> target_prefix
+
+(** {1:info_build_artifacts Build artifacts} *)
+
+(** Those functions changes the target [filename] while conserving both source
+    file and module name metadata.*)
+val cmi: ('a,_) t -> 'a any_target
+val cmt: ('a,_) t -> 'a any_target
+val cmti: ('a,_) t -> 'a any_target
+val annot: ('a,_) t -> 'a any_target
+val obj: ('a,_) t -> 'a any_target
+
+(** The functions below return the filename of build artefact associated to a
+    file. *)
+val cmx_name: _ t -> string
+val cmo_name: _ t -> string
+val obj_name: _ t -> string
+val annot_name: _ t -> string
```

Comment:
For me `name` is not so clear, I prefer `path` (filesystem path) when applicable.

Reply:
On reading the API, it is unclear why some extensions such has `.annot` have both `annot` and `annot_name`, `.cmi` has no `cmi_name` and `.cmx` has no `cmx` function. I would find it more regular to have only functions returning a `('a, string) t` value for those extensions, and then a `target_path : (_, string) t -> path` function to get the output path.